Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py

```diff
@@ -0,0 +1,693 @@
+
+        # setup data structures
+        self.hierarchy = {}  # type: Dict[str, int]
+        self.num_of_packets = 0
+        self.tcp_streams = 0
+        self.udp_streams = 0
+        self.bytes_transmitted = 0
+        self.min_time = float('inf')
+        self.max_time = -float('inf')
+        self.conversations = {}  # type: Dict[tuple, Any]
+        self.flows = {}  # type: Dict[tuple, Any]
+        self.unique_source_ip = set([])  # type: set
+        self.unique_dest_ip = set([])  # type: set
+        self.ips_extracted = set([])  # type: set
+        self.urls_extracted = set([])  # type: set
+        self.emails_extracted = set([])  # type: set
+        self.homemade_extracted = set([])  # type: set
+        self.last_layer = set([])  # type: set
+        self.irc_data = list()  # type: list
+        self.protocol_data = dict()  # type: Dict[str, Any]
+        self.extracted_protocols = extracted_protocols
+        self.homemade_regex = homemade_regex
+        self.unique_ips = unique_ips
+        for protocol in extracted_protocols:
+            self.protocol_data[protocol] = dict()
+
+        # Regex compilation
+        if 'LLMNR' in extracted_protocols:
+            self.llmnr_type = re.compile('Type: (.*)\n')
+            self.llmnr_class = re.compile('Class: (.*)\n')
+            self.llmnr_dict = {}  # type: dict
+
+        if is_reg_extract:
+            self.reg_ip = re.compile(IP_REGEX)
+            self.reg_email = re.compile(EMAIL_REGEX)
+            self.reg_url = re.compile(URL_REGEX)
+
+        if 'HTTP' in extracted_protocols:
+            self.reg_pragma = re.compile(PRAGMA_REGEX)
+
+        if 'ICMP' in extracted_protocols:
+            self.icmp_data = set()  # type: set
+        if 'DNS' in extracted_protocols or 'NETBIOS' in extracted_protocols or 'ICMP' in extracted_protocols:
+            self.reg_type = re.compile(TYPE_REGEX)
+        if 'NETBIOS' in extracted_protocols:
+            self.reg_class = re.compile(CLASS_REGEX)
+        if 'SMB2' in extracted_protocols:
+            self.reg_cmd = re.compile(COMMAND_REGEX)
+        if 'KERBEROS' in extracted_protocols:
+            self.reg_sname = re.compile(SNAME_REGEX)
+            self.kerb_data = set()  # type: set
+        if 'SSH' in extracted_protocols:
+            self.ssh_data = {
+                'ClientProtocols': set(),
+                'ServerProtocols': set(),
+                'KeyExchangeMessageCode': set()
+            }  # type: dict
+            self.reg_message_code = re.compile(MESSAGE_CODE)
+        if 'FTP' in extracted_protocols:
+            self.reg_res_code = re.compile(RESPONSE_CODE)
+        if 'TELNET' in extracted_protocols:
+            self.telnet_data = set()  # type: set
+            self.telnet_commands = set()  # type: set
+        if homemade_regex:
+            self.reg_homemade = re.compile(self.homemade_regex)
+
+    def extract_context_from_packet(self, packet, layers_str: str, is_reg_extract: bool = False) -> None:
```

Comment:
the function is way too long, please split

Reply:
This function has a lot of ifs to handle all the different protocols. Should I add a different function for every protocol? I'm not sure if that would be more concise or faster. What do you think?